File: r565/D.py
from collections import Counter

prime = [True] * 2750132
for i in range(2, 1659):
    if not prime[i]: continue
    for j in range(2*i, len(prime), i):
        prime[j] = False
pr = [i for i,p in enumerate(prime) if p][1:] # 1, 2, 3, 5, 7, ...
#print(pr[:10])
#pos = {p:i-1 for i,p in enumerate(pr) if p}
#print(len(pr), pr[-10:])
#print(pos[2750131])
#print(pos[7])
#exit(1)

input()
b = sorted(map(int, input().split()), reverse=True)
skip = Counter()
p, a = [], [] # prime, a

for v in b:
    if skip[v]:
        skip[v] -= 1
        continue
    prime = True
    if v <= 200000:
        for d in range(2,min(448, v)):
            if v%d: continue
            #print(v, 'is composite, skip', v//d)
            a.append(v) # composite, and not biggest divisor
            skip[v//d] += 1 # skip the biggest divisor later on
            prime = False
            break
    if prime:
        #print(v, 'is prime')
        p.append(v)

skip = Counter() # again
for v in sorted(p):
    if skip[v]:
        skip[v] -= 1
        continue
    a.append(v)
    skip[pr[v]] += 1

print(' '.join(str(v) for v in a))

File: r565/D.cc
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <algorithm>

using namespace std;

typedef long long LL;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    vector<bool> prime(2750132, true);

    for(int i=2; i<1659; i++) {
        if(!prime[i]) continue;
        for(int j=2*i; j<prime.size(); j+=i)
            prime[j] = false;
    }

    vector<int> pr(200001);
    int pos = 0;

    for(int i=1; i<prime.size(); i++)
        if(prime[i]) pr[pos++] = i;

    int n;

    cin >> n;
    vector<int> b(n*2);
    for(int i=0; i<n*2; i++)
        cin >> b[i];
    sort(b.rbegin(), b.rend());

    vector<int> p, a;
    map<int,int> skip;

    //cout << "Go through ";
    //for(int v : b) cout << v << " "; cout << endl;
    for(int v : b) {
        if(skip[v]) {
            skip[v]--;
            continue;
        }
        bool prime = true;
        if(v <= 200000) {
            for(int d = 2; d < min(448, v); d++) {
                if(v%d) continue;
                a.push_back(v); // composite, and not biggest divisor
                skip[v/d]++; // skip the biggest divisor later on
                prime = false;
                break;
            }
        }
        if(prime) p.push_back(v);
    }

    //for(int v : a) cout << v << " "; cout << endl;
    sort(p.begin(), p.end());
    skip.clear();
    //cout << "Iterate ";
    //for(int v : p) cout << v << " "; cout << endl;
    for(int v : p) {
        if(skip[v]) {
            //cout << "Hoy skipping " << v << endl;
            skip[v]--;
            continue;
        }
        a.push_back(v);
        //cout << "Add " << v << " and skip " << pr[v] << endl;
        skip[pr[v]]++;
    }

    for(int v : a) cout << v << " "; cout << endl;
}
